PLFM-9032 - Cache span in patch, builders for immutable Operations

nickgros opened the pull request:
- `Patch`: cache span and return immutable operations
- Add disabled scalability test.
- Updated `Operation` classes
  - All operation classes are now immutable
  - All operation classes now validate that required properties are provided in the constructor
- Added OperationBuilder classes to support creating operations within a patch.

nickgros committed:
PLFM-9032 - Cache span in patch, builders for immutable Operations

- `Patch`: cache span and return immutable operations
- Add disabled scalability test.
- Updated `Operation` classes
  - All operation classes are now immutable
  - All operation classes now validate that required properties are provided in the constructor
- Added OperationBuilder classes to support creating operations within a patch.

nickgros committed:
PLFM-9032

Add missing getters/hash/equals

nickgros committed:
PLFM-9032

Fix compile error

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/Patch.java:
```diff
@@ -38,46 +38,32 @@ public List<Operation<?>> getOperations() {
 
 	public Patch setOperations(List<Operation<?>> operations) {
```
can we remove?

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/Patch.java:
```diff
@@ -38,46 +38,32 @@ public List<Operation<?>> getOperations() {
+	 * It generates the ID and asks the builder to construct the final object.
 	 */
-	public <T extends Operation<T>> T addNewOperation(Class<? extends T> clazz) {
+	public <T extends Operation<T>> T addNewOperation(OperationBuilder<T> builder) {
```
return the ID only.

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/operation/InsertVector.java:
```diff
@@ -4,12 +4,28 @@
+		ValidateArgument.required(operationId, "operationId");
+		ValidateArgument.required(vectorId, "vectorId");
+		ValidateArgument.required(map, "map");
+		if (map.isEmpty()) {
```
can  a client do this?

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/operation/Operation.java:
```diff
@@ -3,31 +3,24 @@
+ * Immutable abstraction shared by all patch operations. See: <a href=
+ * "https://jsonjoy.com/specs/json-crdt-patch/patch-document/patch-structure">patch-structure</a>
  */
 public interface Operation<T> {
```
Can remove generic

john-hill commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/operation/builder/OperationBuilder.java:
```diff
@@ -0,0 +1,16 @@
+import org.sagebionetworks.repo.model.grid.patch.LogicalTimestamp;
+import org.sagebionetworks.repo.model.grid.patch.operation.Operation;
+
+public abstract class OperationBuilder<T extends Operation<T>> {
```
generic

nickgros committed:
PLFM-9032 - Code review changes

- `addNewOperation` returns operation ID only
- Remove unnecessary generic parameters

nickgros commented on lib/lib-grid-db/src/main/java/org/sagebionetworks/grid/db/OperationHandler.java:
```diff
@@ -7,7 +7,7 @@
 import org.sagebionetworks.repo.model.grid.patch.operation.OperationType;
 
-public interface OperationHandler<T extends Operation<T>> {
+public interface OperationHandler<T extends Operation> {
```
This generic parameter cannot be removed because the exact operation implementation is used in the `handleBatch` method

nickgros commented on lib/lib-grid/src/main/java/org/sagebionetworks/repo/model/grid/patch/compact/OperationSerializable.java:
```diff
@@ -5,7 +5,7 @@
 import org.sagebionetworks.repo.model.grid.patch.operation.OperationType;
 
-public interface OperationSerializable<T extends Operation<?>> {
+public interface OperationSerializable<T extends Operation> {
```
Generic parameter cannot be removed.

nickgros committed:
PLFM-9032 - fix erroneous variable replacement
